Quarkus Liquibase-MongoDB Extension is ignoring auth-mechanism and auth-mechanism-properties

### Describe the bug

I am unable to connect to MongoDB Atlas hosted on AWS using quarkus liquibase-mongodb extension to create Collections and Indexes during the application startup.

we are trying to connect with below mentioned properties
```properties
quarkus.mongodb.credentials.auth-source=$external
quarkus.mongodb.credentials.auth-mechanism=MONGODB-AWS
quarkus.mongodb.credentials.password=<AWS_SECRET_ACCESS_KEY>
quarkus.mongodb.credentials.username=<AWS_ACCESS_KEY_ID>
quarkus.mongodb.credentials.auth-mechanism-properties.AWS_SESSION_TOKEN=<AWS_SESSION_TOKEN>
quarkus.mongodb.connection-string=mongodb+srv://<MONGO_DB_URL>?retryWrites=true&w=majority
```

https://www.mongodb.com/docs/manual/reference/connection-string/#mongodb-urioption-urioption.authMechanism

### Expected behavior

Quarkus mongodb properties must get considered and connection to mongodb atlas must be established

### Actual behavior

```posh
Wrong Auth-Mechanism is being used "SCRAM-SHA-1" instead of "MONGODB-AWS"

31-08-2023 18:37:12,942 [main] prozessid:  transaktionsid:  ERROR [io.quarkus.runtime.Application] Failed to start application (with profile [prod]): java.lang.RuntimeException: Failed to start quarkus
--
at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
at io.quarkus.runtime.Application.start(Application.java:101)
at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:111)
at io.quarkus.runtime.Quarkus.run(Quarkus.java:71)
at io.quarkus.runtime.Quarkus.run(Quarkus.java:44)
at io.quarkus.runtime.Quarkus.run(Quarkus.java:124)
at io.quarkus.runner.GeneratedMain.main(Unknown Source)
Caused by: java.lang.IllegalStateException: Error starting Liquibase
at io.quarkus.liquibase.mongodb.runtime.LiquibaseMongodbRecorder.doStartActions(LiquibaseMongodbRecorder.java:73)
at io.quarkus.deployment.steps.LiquibaseMongodbProcessor$startLiquibase1767457900.deploy_0(Unknown Source)
at io.quarkus.deployment.steps.LiquibaseMongodbProcessor$startLiquibase1767457900.deploy(Unknown Source)
... 7 more
Caused by: liquibase.exception.DatabaseException: Could not query for list
at liquibase.nosql.executor.NoSqlExecutor.queryForList(NoSqlExecutor.java:141)
at liquibase.nosql.executor.NoSqlExecutor.queryForList(NoSqlExecutor.java:131)
at liquibase.ext.mongodb.changelog.MongoHistoryService.queryRanChangeSets(MongoHistoryService.java:128)
at liquibase.nosql.changelog.AbstractNoSqlHistoryService.getRanChangeSets(AbstractNoSqlHistoryService.java:149)
at liquibase.database.AbstractJdbcDatabase.getRanChangeSetList(AbstractJdbcDatabase.java:1139)
at liquibase.changelog.DatabaseChangeLog.validate(DatabaseChangeLog.java:345)
at liquibase.changelog.DatabaseChangeLog.validate(DatabaseChangeLog.java:330)
at liquibase.Liquibase.validate(Liquibase.java:2577)
at io.quarkus.liquibase.mongodb.runtime.LiquibaseMongodbRecorder.doStartActions(LiquibaseMongodbRecorder.java:63)
... 9 more
Caused by: com.mongodb.MongoSecurityException: Exception authenticating MongoCredential{mechanism=**SCRAM-SHA-1**, userName=<AWS_ACCESS_KEY_ID> source='$external', password=<hidden>, mechanismProperties=<hidden>}
at com.mongodb.internal.connection.SaslAuthenticator.wrapException(SaslAuthenticator.java:270)
at com.mongodb.internal.connection.SaslAuthenticator.getNextSaslResponse(SaslAuthenticator.java:133)
at com.mongodb.internal.connection.SaslAuthenticator.lambda$authenticate$0(SaslAuthenticator.java:63)
at com.mongodb.internal.connection.SaslAuthenticator.doAsSubject(SaslAuthenticator.java:277)
at com.mongodb.internal.connection.SaslAuthenticator.authenticate(SaslAuthenticator.java:59)
at com.mongodb.internal.connection.DefaultAuthenticator.authenticate(DefaultAuthenticator.java:57)
at com.mongodb.internal.connection.InternalStreamConnectionInitializer.authenticate(InternalStreamConnectionInitializer.java:205)
at com.mongodb.internal.connection.InternalStreamConnectionInitializer.finishHandshake(InternalStreamConnectionInitializer.java:85)
at com.mongodb.internal.connection.InternalStreamConnection.open(InternalStreamConnection.java:209)
at com.mongodb.internal.connection.UsageTrackingInternalConnection.open(UsageTrackingInternalConnection.java:55)
at com.mongodb.internal.connection.DefaultConnectionPool$PooledConnection.open(DefaultConnectionPool.java:534)
at com.mongodb.internal.connection.DefaultConnectionPool$OpenConcurrencyLimiter.openWithConcurrencyLimit(DefaultConnectionPool.java:895)
at com.mongodb.internal.connection.DefaultConnectionPool$OpenConcurrencyLimiter.openOrGetAvailable(DefaultConnectionPool.java:836)
at com.mongodb.internal.connection.DefaultConnectionPool.get(DefaultConnectionPool.java:177)
at com.mongodb.internal.connection.DefaultConnectionPool.get(DefaultConnectionPool.java:166)
at com.mongodb.internal.connection.DefaultServer.getConnection(DefaultServer.java:97)
at com.mongodb.internal.binding.ClusterBinding$ClusterBindingConnectionSource.getConnection(ClusterBinding.java:173)
at com.mongodb.client.internal.ClientSessionBinding$SessionBindingConnectionSource.getConnection(ClientSessionBinding.java:193)
at com.mongodb.internal.operation.OperationHelper.withSuppliedResource(OperationHelper.java:375)
at com.mongodb.internal.operation.OperationHelper.lambda$withSourceAndConnection$1(OperationHelper.java:357)
at com.mongodb.internal.operation.OperationHelper.withSuppliedResource(OperationHelper.java:383)
at com.mongodb.internal.operation.OperationHelper.withSourceAndConnection(OperationHelper.java:356)
at com.mongodb.internal.operation.FindOperation.lambda$execute$2(FindOperation.java:323)
at com.mongodb.internal.async.function.RetryingSyncSupplier.get(RetryingSyncSupplier.java:67)
at com.mongodb.internal.operation.FindOperation.execute(FindOperation.java:334)
at com.mongodb.internal.operation.FindOperation.execute(FindOperation.java:73)
at com.mongodb.client.internal.MongoClientDelegate$DelegateOperationExecutor.execute(MongoClientDelegate.java:185)
at com.mongodb.client.internal.MongoIterableImpl.execute(MongoIterableImpl.java:130)
at com.mongodb.client.internal.MongoIterableImpl.iterator(MongoIterableImpl.java:90)
at com.mongodb.client.internal.MongoIterableImpl.forEach(MongoIterableImpl.java:116)
at com.mongodb.client.internal.MongoIterableImpl.into(MongoIterableImpl.java:125)
at liquibase.ext.mongodb.statement.FindAllStatement.queryForList(FindAllStatement.java:79)
at liquibase.ext.mongodb.statement.FindAllStatement.queryForList(FindAllStatement.java:36)
at liquibase.nosql.executor.NoSqlExecutor.queryForList(NoSqlExecutor.java:139)
... 17 more
Caused by: com.mongodb.MongoCommandException: Command failed with error 334 (MechanismUnavailable): 'Unsupported mechanism 'SCRAM-SHA-1' on authentication database '$external'' on server <HOST>:<PORT>. The full response is {     "ok": 0,     "errmsg": "Unsupported mechanism 'SCRAM-SHA-1' on authentication database '$external'",     "code": 334,     "codeName": "MechanismUnavailable",     "$clusterTime": {         "clusterTime": {             "$timestamp": {                 "t": 1693499825,                 "i": 1             }         },         "signature": {             "hash": {                 "$binary": {                     "base64": "E6n2PfelrMfS1oe7+DJhM79ZGcc=",                     "subType": "00"                 }             },             "keyId": 7247132067840917508         }     },     "operationTime": {         "$timestamp": {             "t": 1693499825,             "i": 1         }     } }
at com.mongodb.internal.connection.ProtocolHelper.getCommandFailureException(ProtocolHelper.java:205)
at com.mongodb.internal.connection.InternalStreamConnection.receiveCommandMessageResponse(InternalStreamConnection.java:443)
at com.mongodb.internal.connection.InternalStreamConnection.sendAndReceive(InternalStreamConnection.java:365)
at com.mongodb.internal.connection.CommandHelper.sendAndReceive(CommandHelper.java:102)
at com.mongodb.internal.connection.CommandHelper.executeCommand(CommandHelper.java:49)
at com.mongodb.internal.connection.SaslAuthenticator.sendSaslStart(SaslAuthenticator.java:224)
at com.mongodb.internal.connection.SaslAuthenticator.getNextSaslResponse(SaslAuthenticator.java:131)
... 49 more
```

### How to Reproduce?

Use quarkus mongodb properties and quakus-liquibase-mongodb extension with Mongodb Atlas 

```properties
quarkus.mongodb.credentials.auth-source=$external
quarkus.mongodb.credentials.auth-mechanism=MONGODB-AWS
quarkus.mongodb.credentials.password=<AWS_SECRET_ACCESS_KEY>
quarkus.mongodb.credentials.username=<AWS_ACCESS_KEY_ID>
quarkus.mongodb.credentials.auth-mechanism-properties.AWS_SESSION_TOKEN=<AWS_SESSION_TOKEN>
quarkus.mongodb.connection-string=mongodb+srv://<MONGO_DB_URL>?retryWrites=true&w=majority
quarkus.mongodb.database=<DatabaseName>
```

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

17

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

3.2.4.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Maven 3.8.6

### Additional information

_No response_

--- a/docs/src/main/asciidoc/liquibase-mongodb.adoc
+++ b/docs/src/main/asciidoc/liquibase-mongodb.adoc
@@ -81,6 +81,8 @@ quarkus.liquibase-mongodb.migrate-at-start=true
 # quarkus.liquibase-mongodb.default-schema-name=DefaultSchema
 ----
 
+NOTE: Liquibase MongoDB is configured using a connection string, we do our best to craft a connection string that matches the MongoDB client configuration but if some configuration properties are not working you may consider adding them directly into the `quarkus.mongodb.connection-string` config property.
+
 Add a changeLog file to the default folder following the Liquibase naming conventions: `{change-log}`
 YAML, JSON and XML formats are supported for the changeLog.
 

--- a/extensions/liquibase-mongodb/runtime/src/main/java/io/quarkus/liquibase/mongodb/LiquibaseMongodbFactory.java
+++ b/extensions/liquibase-mongodb/runtime/src/main/java/io/quarkus/liquibase/mongodb/LiquibaseMongodbFactory.java
@@ -37,6 +37,9 @@ public class LiquibaseMongodbFactory {
                 Thread.currentThread().getContextClassLoader())) {
             String connectionString = this.mongoClientConfig.connectionString.orElse("mongodb://localhost:27017");
 
+            // Every MongoDB client configuration must be added to the connection string, we didn't add all as it would be too much to support.
+            // For reference, all connections string options can be found here: https://www.mongodb.com/docs/manual/reference/connection-string/#connection-string-options.
+
             Matcher matcher = HAS_DB.matcher(connectionString);
             if (!matcher.matches() || matcher.group("db") == null || matcher.group("db").isEmpty()) {
                 connectionString = matcher.replaceFirst(
@@ -50,6 +53,17 @@ public class LiquibaseMongodbFactory {
                 boolean alreadyHasQueryParams = connectionString.contains("?");
                 connectionString += (alreadyHasQueryParams ? "&" : "?") + "authSource="
                         + mongoClientConfig.credentials.authSource.get();
+            }
+            if (mongoClientConfig.credentials.authMechanism.isPresent()) {
+                boolean alreadyHasQueryParams = connectionString.contains("?");
+                connectionString += (alreadyHasQueryParams ? "&" : "?") + "authMechanism="
+                        + mongoClientConfig.credentials.authMechanism.get();
+            }
+            if (!mongoClientConfig.credentials.authMechanismProperties.isEmpty()) {
+                boolean alreadyHasQueryParams = connectionString.contains("?");
+                connectionString += (alreadyHasQueryParams ? "&" : "?") + "authMechanismProperties="
+                        + mongoClientConfig.credentials.authMechanismProperties.entrySet().stream()
+                                .map(prop -> prop.getKey() + ":" + prop.getValue()).collect(Collectors.joining(","));
             }
 
             Database database = DatabaseFactory.getInstance().openDatabase(connectionString,

